src/compose.ts

```diff
@@ -346,3 +366,99 @@ function normalizeKeyValuePairs(
 		.fromPairs()
 		.value();
 }
+
+/**
+ * Parses a composition and returns a list of image descriptors that can
+ * be used to pull or build a service image. The given composition version
+ * must be equal to `DEFAULT_SCHEMA_VERSION`, or an exception is thrown.
+ * Normalise the composition before passing it to this function.
+ */
+async function expandContext(
+	c: Composition,
+	fileResolverCb: (path: string) => Promise<ReadStream>,
+): Promise<Composition> {
+	const cachedEnvironmentFiles: Dict<Dict<string>> = {};
+
+	// Apply all read env_files content to the services referncing the env_files
+	for (const service of Object.values(c.services)) {
+		let envFilePaths = service.env_file;
+		if (envFilePaths !== undefined) {
+			service.environment = service.environment ?? {};
+			if (!Array.isArray(envFilePaths)) {
+				envFilePaths = [envFilePaths];
+			}
+			for (const envFilePath of envFilePaths) {
+				const validatedPath = validateAndNormalizeFilePath(envFilePath);
+				if (!(validatedPath in cachedEnvironmentFiles)) {
+					cachedEnvironmentFiles[validatedPath] = await readAndExpandEnvFile(
```

Comment:
My personal preference would be split this single loop into two operations. 
1. Loading and caching env files
2. Updating the service's environment